tests/unit/products.t

```diff
@@ -20,6 +20,16 @@ is(normalize_code(' just a simple test 036000291452 here we go '),
 	'0036000291452', 'should add leading 0 to cleaned valid UPC12');
 is(normalize_code(' just a simple test 036000291455 here we go '),
 	'036000291455', 'should not add leading 0 to cleaned invalid UPC12');
+is(normalize_code('0104044782317112'), '4044782317112', 'should reduce GS1 AI unbracketed string to GTIN');
+is(normalize_code('(01)04044782317112(17)270101'), '4044782317112', 'should reduce GS1 AI bracketed string to GTIN');
+is(normalize_code('^010404478231711217270101'),
+	'4044782317112', 'should reduce GS1 AI unbracketed string with ^ as FNC1 to GTIN');
+is(normalize_code("\x{001d}010404478231711217270101"),
+	'4044782317112', 'should reduce GS1 AI unbracketed string with original FNC1 to GTIN');
+is(normalize_code("\x{241d}010404478231711217270101"),
+	'4044782317112', 'should reduce GS1 AI unbracketed string with GS as FNC1 to GTIN');
+is(normalize_code('https://id.gs1.org/01/04044782317112/22/2A'),
```

Comment:
Awesome! Could you add a few more tests?
e.g. from GS1 implementation guide:
https://dalgiardino.com/01/09506000134376/10/ABC/21/123456?17=211200
https://dalgiardino.com/01/09506000134369

Maybe some QR code urls that are not GS1 Digital Links, like https://world.openfoodfacts.org/

Reply:
Yup, I was adding more tests from the GS1 guide when I noticed that DataMatrix doesn't work with ZBar. This feature probably doesn't have that much use before that, because GS1 DataMatrix seems to be *the* major format, as it's already being used.

Reply:
Added a first test for GS1 DataMatrix. I'll have a look at some additional test cases in the upcoming days.

Reply:
I added a handful of more tests. There's one case I'm not quite sure how you want to handle it: Even if the new GS1 parser isn't used `normalize_code('http://spam.zip/420')` would return `420`, because of `$code =~ s/\D//g;`. Strictly speaking, keeping that case is probably better for backwards compatibility, but it's probably not a valid GS1 Data Link. 🤷

Reply:
@stephanegigandet I tried adding that restriction in baf76d5037b0e18e181486584cfe9e4a3b016ca4, but this change causes some integration tests to fail: https://github.com/openfoodfacts/openfoodfacts-server/actions/runs/6150811346/job/16689935236?pr=8911#step:4:16384 and https://github.com/openfoodfacts/openfoodfacts-server/actions/runs/6150811346/job/16689935236?pr=8911#step:4:19025

By rejecting shorter codes during normalization, would you rather have the limit be removed or the integration tests adjusted?

Reply:
BTW: The API checks for `^\d{4,24}$`, so maybe that's a better indicator (or at least more consistent with the current behaviour)?

Reply:
@hangy It would be good to be consistent, I guess we can keep 4 to 24, and adjust the tests to put a few more digits in the test barcode :)

Reply:
Turns out this is a bit more complicated, because `normalize_code`, in the original version, removes everything but digits, so that a theoretically invalid `http://spam.example/id/01/123` would be "normalized" to `0123`. I think I'll postpone this until #8926.